core/base/compactTriangulation/CompactTriangulation.h

```diff
@@ -0,0 +1,4133 @@
+/// to be subdivided before using TopoCluster, i.e., there has to be a scalar
+/// field named "ttkCompactTriangulationIndex" to denote the cluster index of
+/// each vertex. Note Topocluster will reindex the simplices based on the
+/// clustering input array.
+/// \b Related \b publications \n
+/// "TopoCluster: A Localized Data Structure for Topology-based Visualization"
+/// Guoxi Liu, Federico Iuricich, Riccardo Fellegara, and Leila De Floriani
+/// IEEE Transactions on Visualization and Computer Graphics, 2021.
+///
+/// \sa ttk::Triangulation
+/// \sa ttk::CompactTriangulationPreconditioning
+
+#pragma once
+
+// base code includes
+#include <AbstractTriangulation.h>
+#include <CellArray.h>
+#include <FlatJaggedArray.h>
+#include <algorithm>
+#include <boost/functional/hash.hpp>
+#include <boost/unordered_map.hpp>
+#include <boost/unordered_set.hpp>
+#include <list>
+
+#define EDGE_ID 1
+#define TRIANGLE_ID 2
+
+namespace ttk {
+
+  class ImplicitCluster {
+  private:
+    /* components */
+    SimplexId nid;
+    std::vector<std::array<SimplexId, 2>> internalEdgeList_;
+    std::vector<std::array<SimplexId, 3>> internalTriangleList_;
+    boost::unordered_map<std::array<SimplexId, 2>, SimplexId> internalEdgeMap_;
+    boost::unordered_map<std::array<SimplexId, 2>, SimplexId> externalEdgeMap_;
+    boost::unordered_map<std::array<SimplexId, 3>, SimplexId>
+      internalTriangleMap_;
+    boost::unordered_map<std::array<SimplexId, 3>, SimplexId>
+      externalTriangleMap_;
+    /* boundary cells */
+    std::vector<bool> boundaryVertices_;
+    std::vector<bool> boundaryEdges_;
+    std::vector<bool> boundaryTriangles_;
+    /* vertex relationships */
+    FlatJaggedArray vertexEdges_;
+    FlatJaggedArray vertexLinks_;
+    FlatJaggedArray vertexNeighbors_;
+    FlatJaggedArray vertexStars_;
+    FlatJaggedArray vertexTriangles_;
+    /* edge relationships */
+    // edgeVertex relation can be extracted from internal edge list
+    FlatJaggedArray edgeLinks_;
+    FlatJaggedArray edgeStars_;
+    FlatJaggedArray edgeTriangles_;
+    /* triangle relationships */
+    // triangleVertex relation can be extracted from internal triangle list
+    std::vector<std::array<SimplexId, 3>> triangleEdges_;
+    FlatJaggedArray triangleLinks_;
+    FlatJaggedArray triangleStars_;
+    /* cell relationships */
+    std::vector<std::array<SimplexId, 6>> tetraEdges_;
+    FlatJaggedArray cellNeighbors_;
+    std::vector<std::array<SimplexId, 4>> tetraTriangles_;
+
+  public:
+    ImplicitCluster() {
+    }
+    ImplicitCluster(SimplexId id) : nid(id) {
+    }
+    ~ImplicitCluster() {
+    }
+
+    friend class CompactTriangulation;
+  };
+
+  class CompactTriangulation final : public AbstractTriangulation {
+
+  public:
+    CompactTriangulation();
+
+    ~CompactTriangulation();
+
+    /**
+     * Set up vertices from the input.
+     */
+    int setInputPoints(const SimplexId &pointNumber,
```

Comment:
This method should either be marked `inline` or moved to the .cpp source file to enforce the One-Definition Rule (ODR). Moving to the .cpp might also have a positive impact on the build times.